File: java/unittests/queue/LinkedList.java
/*
 * NAME: LinkedList
 * AUTHOR: Tanaka Chitete
 * PURPOSE: Implement a generic Linked List
 * CREATION: 31/08/2020
 * LAST MODIFICATION: 09/11/2020
 */

import java.io.Serializable;
import java.util.*;

public class LinkedList<T> extends AbstractSequentialList<T> implements Serializable {
    // NESTED CLASSES

    /*
     * NAME: Node
     * CREATOR: Tanaka Chitete
     * PURPOSE: Implement nodes for LinkedList
     * CREATION: 31/08/2020
     * LAST MODIFICATION: 03/11/2020
     */

    private class Node<A> implements Serializable {
        // PRIVATE CLASS FIELDS
    
        private A element;
        private Node<A> previousNode;
        private Node<A> nextNode;
    
        // CONSTRUCTORS
    
        /*
         * ALTERNATE CONSTRUCTOR
         * IMPORT(S): inElement (Object)
         * EXPORT(S): Address of new Node
         * PURPOSE: Create new Node in alternate state
         * CREATION: 31/08/2020
         * LAST MODIFICATION: 03/11/2020
         */
    
        private Node(A inElement) {
            element = inElement;
            previousNode = null;
            nextNode = null;
        }
    }

    // PRIVATE CLASS FIELDS

    private Node<T> firstNode;
    private Node<T> lastNode;
    private int count;

    // CONSTRUCTORS

    /*
     * DEFAULT CONSTRUCTOR
     * IMPORT(S): NONE
     * EXPORT(S): Address of new LinkedList
     * PURPOSE: Create new LinkedList in default state
     * CREATION: 31/08/2020
     * LAST MODIFICATION: 3/11/2020
     */

    public LinkedList() {
        firstNode = null;
        lastNode = null;
        count = 0;
    }

    // SETTERS (MUTATORS)

    public void insertFirst(T newFirstNodeElement) {
        Node<T> newFirstNode = new Node<T>(newFirstNodeElement);

        if (isEmpty()) {
            firstNode = newFirstNode;
            lastNode = newFirstNode;
        }
        else {
            newFirstNode.nextNode = firstNode;
            firstNode = newFirstNode;
            count++;
        }
    }

    public void insertLast(T newLastNodeElement) {
        Node<T> newLastNode = new Node<T>(newLastNodeElement);

        if (isEmpty()) {
            lastNode = newLastNode;
            firstNode = newLastNode;
        }
        else {
            lastNode.nextNode = newLastNode;
            lastNode = newLastNode;
            count++;
        }
    }

    // GETTERS (ACCESSORS)

    public T peekFirst() {
        T peekedFirstNodeElement;

        if (isEmpty()) {
            throw new IllegalStateException("Cannot peek at the first node element of an empty " + 
                "linked list");
        }
        else {
            peekedFirstNodeElement = firstNode.element;
        }

        return peekedFirstNodeElement;
    }

    public T peekLast() {
        T peekedLastNodeElement;

        if (isEmpty()) {
            throw new IllegalStateException("Cannot peek at the last node element of an empty " + 
                "linked list");
        }
        else {
            peekedLastNodeElement = lastNode.element;
        }

        return peekedLastNodeElement;
    }

    public T removeFirst() {
        T removedFirstNodeElement;

        if (isEmpty()) {
            throw new IllegalStateException("Cannot remove the first node element of an empty " + 
                "linked list");
        }
        else {
            removedFirstNodeElement = firstNode.element;
            firstNode = firstNode.nextNode;
            count--;
        }

        return removedFirstNodeElement;
    }

    public T removeLast() {
        T removedLastNodeElement;

        if (isEmpty()) {
            throw new IllegalStateException("Cannot remove the last node element of an empty " + 
                "linked list");
        }
        else if (firstNode.nextNode == null) {
            removedLastNodeElement = firstNode.element;
            firstNode = null;
        }
        else {
            removedLastNodeElement = lastNode.element;
            lastNode = lastNode.previousNode;
            lastNode.nextNode = null;
            count--;
        }

        return removedLastNodeElement;
    }

    // OPERATORS

    public boolean isEmpty() {
        boolean isEmpty;

        if (firstNode == null) {
            isEmpty = true;
        }
        else {
            isEmpty = false;
        }

        return isEmpty;
    }

    @Override
    public int size() {
        return count;
    }   
    
    // ITERATOR

    @Override 
    public ListIterator<T> listIterator(int index) {
        throw new UnsupportedOperationException();
    }
    
    public Iterator<T> iterator() {
        // Hooks iterator to this LinkedList object
        return new DSALinkedListIterator(this);
    }

    private class DSALinkedListIterator implements Iterator<T> {
        private Node<T> iterNext;

        public DSALinkedListIterator(LinkedList<T> inDSALinkedList) {
            iterNext = inDSALinkedList.firstNode;
        }

        @Override
        public boolean hasNext() {
            boolean hasNext;

            if (iterNext == null) {
                hasNext = false;
            }
            else {
                hasNext = true;
            }

            return hasNext;
        }

        @Override
        public T next() {
            T element;

            if (iterNext == null) {
                element = null;
            }
            else {
                element = iterNext.element;
                iterNext = iterNext.nextNode;
            }

            return element;
        }

        public void remove() {
            throw new UnsupportedOperationException();
        }
    }
}

File: python/unittests/linkedlist/LinkedList.py
"""
NAME: LinkedList
AUTHOR: Tanaka Chitete
PURPOSE: Implement a Linked List
CREATION: 27/02/2021
LAST MODIFICATION: 01/03/2021
"""

from LinkedListNode import LinkedListNode

class LinkedList:
    # CONSTRUCTORS

    """
    DEFAULT CONSTRUCTOR
    IMPORT(S): NONE
    EXPORT(S): Address of new LinkedList
    PURPOSE: Make new LinkedList in default state
    CREATION: 28/02/2021
    LAST MODIFICATION: 28/02/2021
    """

    def __init__(self):
        self.__first = None
        self.__last = None
        self.__count = 0


    # SETTERS (MUTATORS)

    def insertFirst(self, newFirstElement):
        newFirst = LinkedListNode(newFirstElement)

        if self.isEmpty():
            self.__first = newFirst
            self.__last = newFirst
        else:
            newFirst.next = self.__first
            self.__first = newFirst
        self.__count += 1


    def insertLast(self, newLastElement):
        newLast = LinkedListNode(newLastElement)

        if self.isEmpty():
            self.__last = newLast
            self.__first = newLast
        else:
            self.__last.next = newLast
            self.__last = newLast
        self.__count += 1


    # GETTERS (ACCESSORS)

    def peekFirst(self):
        if self.isEmpty():
            raise ValueError("Cannot peek at the first node of an empty linked list")
        else:
            return self.__first.element


    def peekLast(self):
        if self.isEmpty():
            raise ValueError("Cannot peek at the last node of an empty linked list")
        else:
            return self.__last.element


    def removeFirst(self):
        if self.isEmpty():
            raise ValueError("Cannot remove the first node of an empty linked list")
        else:
            removedFirstElement = self.__first.element
            self.__first = self.__first.next
            self.__count -= 1
            return removedFirstElement


    def removeLast(self):
        if self.isEmpty():
            raise ValueError("Cannot remove the last node of an empty linked list")
        elif self.__first.next == None:
            removedLastElement = self.__first.element
            self.__first = None
            self.__count -= 1
            return removedLastElement
        else:
            removedLastElement = self.__last.element
            self.__last = self.__last.previous
            self.__last.next = None
            self.__count -= 1
            return removedLastElement


    # OPERATORS

    def isEmpty(self):
        return self.__first is None


    def __len__(self):
        return self.__count


    # ITERATOR

    def __iter__(self):
        return Iterator(self)
    

    class Iterator:
        def __init__(self, inLinkedList):
            self.next = inLinkedList.__first

        def __next__(self):
            element = None
            if self.next is not None:
                element = self.next.element
                self.next = self.next.next
            return element